### ISSUE TITLE
QK norm as a fix to Pixel DiT explosions.

### ISSUE THREAD

**Issue by theraptorlake** [NVlabs/PixelDiT#14]
While researching on Pixel DiT, to fix the explosion problem, a better solution is to normalize QK with layer/RMS norm instead of using post-norm(which hurt backprop gradients and causes issues), this allows Pixel DiT to stay pre-norm at all stages, QK norm was proposed by google as a solution to make attention more stable. QK norm generally also shows improved performance and quality.

@yeates @wxiong-vision

**Comment by wxiong-vision** [comment]
Thanks for the constructive suggestion! Yeah, we have tried using kq-norm, but it did not resolve the loss spike issue. I wonder if you have tried that on your side, and how does it work? Could you show some training curves?

**Comment by theraptorlake** [comment]
@wxiong-vision Yes, layer norm specifically without affine params worked for me. To solve the NaN explosions, as you mentioned that pixel wise AdaLN sometimes grows too large and when it enters attention it explodes. Specifically to solve this, I used layer norm and disabled affine scale and shift which are built in. I preferred layer norm because RMS norm sometimes suffers from signal quantization risk, layer norm is a bit superior to RMS norm in that terms. Here's my QK-norm implementation. I also preferred Layer norm instead of RMS norm because it prevents the mean to drift away, it's generally better to use that instead of RMS norm especially in sensitive scenarios according to me.

I am really sorry I don't have a training curve right now, but in my earlier experiments, I was training PixelDiT from scratch and at the 276K optim step, it exploded into NaNs. To fix this I implemented this exact QK norm and retrained, it obviously needed a few steps to adjust to the new attn norm, but I didn't see an explosion after that. It was a total pre-norm network, without any post norms. 

```
import torch

import torch.nn as nn
import torch.nn.functional as F

from core.position_embedding.rotary import RotaryPositionEmbedding2D

class MultiHeadAttention(nn.Module):
    def __init__(self, d_attn: int, n_heads: int):
        super().__init__()

        if d_attn % n_heads != 0:
            raise ValueError('d_attn must be divisible by n_heads.')

        self.n_heads = n_heads
        self.d_head = d_attn // n_heads

        self.rope = RotaryPositionEmbedding2D(self.d_head)

        self.q_proj = nn.Linear(d_attn, d_attn, bias=False)
        self.k_proj = nn.Linear(d_attn, d_attn, bias=False)
        self.v_proj = nn.Linear(d_attn, d_attn, bias=False)

        self.q_norm = nn.LayerNorm(self.d_head, elementwise_affine=False)
        self.k_norm = nn.LayerNorm(self.d_head, elementwise_affine=False)

        self.out_proj = nn.Linear(d_attn, d_attn, bias=False)

    def forward(self, x: torch.Tensor, spat_shape: tuple[int, int]) -> torch.Tensor:
        B, T, C = x.shape

        q = self.q_proj(x).reshape(B, T, self.n_heads, self.d_head).transpose(1, 2)
        k = self.k_proj(x).reshape(B, T, self.n_heads, self.d_head).transpose(1, 2)
        v = self.v_proj(x).reshape(B, T, self.n_heads, self.d_head).transpose(1, 2)

        q = self.q_norm(q)
        k = self.k_norm(k)

        q, k = self.rope(q, k, spat_shape)

        x = F.scaled_dot_product_attention(q, k, v, is_causal=False)
        x = x.transpose(1, 2).contiguous().reshape(B, T, C)

        x = self.out_proj(x)

        return x

```
